File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/integrasjoner/enhetsregisteret/EnhetsregisteretClient.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.enhetsregisteret;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.InstitusjonellSektorkode;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.Næringskode5Siffer;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.Orgnr;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.OverordnetEnhet;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.Underenhet;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.Virksomhet;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.bransjeprogram.Bransjeprogram;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Component;
import org.springframework.web.client.HttpServerErrorException;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestClientResponseException;
import org.springframework.web.client.RestTemplate;

@Component
public class EnhetsregisteretClient {

  private final RestTemplate restTemplate;
  private final String enhetsregisteretUrl;

  private static final ObjectMapper objectMapper = new ObjectMapper();

  public EnhetsregisteretClient(
      RestTemplate restTemplate, @Value("${enhetsregisteret.url}") String enhetsregisteretUrl) {
    this.restTemplate = restTemplate;
    this.enhetsregisteretUrl = enhetsregisteretUrl;
  }

  public OverordnetEnhet hentInformasjonOmEnhet(Orgnr orgnrTilEnhet) {
    String url = enhetsregisteretUrl + "enheter/" + orgnrTilEnhet.getVerdi();

    try {
      String respons = restTemplate.getForObject(url, String.class);
      OverordnetEnhet overordnetEnhet = mapTilEnhet(respons);
      validerReturnertOrgnr(orgnrTilEnhet, overordnetEnhet.getOrgnr());
      return overordnetEnhet;
    } catch (RestClientException e) {
      throw new EnhetsregisteretException("Feil ved kall til Enhetsregisteret", e);
    }
  }

  public Underenhet hentInformasjonOmUnderenhet(Orgnr orgnrTilUnderenhet) {
    try {
      String url = enhetsregisteretUrl + "underenheter/" + orgnrTilUnderenhet.getVerdi();
      String respons = restTemplate.getForObject(url, String.class);
      Underenhet underenhet = mapTilUnderenhet(respons);
      validerReturnertOrgnr(orgnrTilUnderenhet, underenhet.getOrgnr());
      return underenhet;
    } catch (HttpServerErrorException hsee) {
      throw new EnhetsregisteretIkkeTilgjengeligException("Enhetsregisteret svarer ikke", hsee);
    } catch (RestClientException e) {
      throw new EnhetsregisteretException("Feil ved kall til Enhetsregisteret", e);
    }
  }

  private OverordnetEnhet mapTilEnhet(String jsonResponseFraEnhetsregisteret) {
    try {
      JsonNode enhetJson = objectMapper.readTree(jsonResponseFraEnhetsregisteret);
      JsonNode næringskodeJson = enhetJson.get("naeringskode1");
      JsonNode sektorJson = enhetJson.get("institusjonellSektorkode");

      return new OverordnetEnhet(
          new Orgnr(enhetJson.get("organisasjonsnummer").textValue()),
          enhetJson.get("navn").textValue(),
          objectMapper.treeToValue(næringskodeJson, Næringskode5Siffer.class),
          objectMapper.treeToValue(sektorJson, InstitusjonellSektorkode.class),
          enhetJson.get("antallAnsatte").intValue());

    } catch (IOException | NullPointerException | IllegalArgumentException e) {
      throw new EnhetsregisteretMappingException(
          "Feil ved kall til Enhetsregisteret. Kunne ikke parse respons.", e);
    }
  }

  private void validerReturnertOrgnr(Orgnr opprinneligOrgnr, Orgnr returnertOrgnr) {
    if (!opprinneligOrgnr.equals(returnertOrgnr)) {
      throw new OrgnrEksistererIkkeException(
          "Orgnr hentet fra Enhetsregisteret samsvarer ikke med det medsendte orgnr. Request: "
              + opprinneligOrgnr.getVerdi()
              + ", response: "
              + returnertOrgnr.getVerdi());
    }
  }

  private Underenhet mapTilUnderenhet(String jsonResponseFraEnhetsregisteret) {
    try {
      JsonNode enhetJson = objectMapper.readTree(jsonResponseFraEnhetsregisteret);

      Næringskode5Siffer primærnæringskode =
          Bransjeprogram.velgPrimærnæringskode(
              List.of("naeringskode1", "naeringskode2", "naeringskode3").stream()
                  .map(enhetJson::get)
                  .filter(Objects::nonNull)
                  .map(json -> objectMapper.convertValue(json, Næringskode5Siffer.class))
                  .collect(Collectors.toList()));

      return new Underenhet(
          new Orgnr(enhetJson.get("organisasjonsnummer").textValue()),
          new Orgnr(enhetJson.get("overordnetEnhet").textValue()),
          enhetJson.get("navn").textValue(),
          primærnæringskode,
          enhetJson.get("antallAnsatte").intValue());

    } catch (IOException | NullPointerException e) {
      throw new EnhetsregisteretMappingException(
          "Feil ved kall til Enhetsregisteret. Kunne ikke parse respons.", e);
    }
  }

  }

  public HttpStatusCode healthcheck() {
    try {
      ResponseEntity<String> response =
          restTemplate.exchange(
              enhetsregisteretUrl, HttpMethod.GET, HttpEntity.EMPTY, String.class);
      return response.getStatusCode();
    } catch (RestClientResponseException e) {
      return HttpStatus.valueOf(e.getStatusCode().value());
    }
  }
}

File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/integrasjoner/enhetsregisteret/EnhetsregisteretClient.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.enhetsregisteret

import arrow.core.Either
import arrow.core.left
import arrow.core.right
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.*
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.config.ProxyWebClient
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.enhetsregisteret.json.OverordnetEnhetResponseJson
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.enhetsregisteret.json.UnderenhetResponseJson
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.core.codec.DecodingException
import org.springframework.http.*
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClientResponseException
import reactor.util.retry.Retry

@Component
open class EnhetsregisteretClient(
    private val webClient: ProxyWebClient,
    @param:Value("\${enhetsregisteret.url}") private val enhetsregisteretUrl: String
) {
    private val log: Logger = LoggerFactory.getLogger(this::class.java)
    private val retrySpec = Retry.max(2).filter {
        it !is DecodingException
    }.onRetryExhaustedThrow { _, signal -> signal.failure() }

    open fun hentEnhet(orgnrTilEnhet: Orgnr): Either<HentEnhetFeil, OverordnetEnhet> {
        val url = "${enhetsregisteretUrl}/enheter/{orgnr}"
        return try {

            val overordnetEnhet = webClient.get()
                .uri(url, orgnrTilEnhet.verdi)
                .retrieve()
                .bodyToMono(OverordnetEnhetResponseJson::class.java)
                .retryWhen(retrySpec)
                .block()!!
                .toDomain()
            if (!validerReturnertOrgnr(orgnrTilEnhet, overordnetEnhet.orgnr)) {
                return HentEnhetFeil.OrgnrMatcherIkke.left()
            }
            overordnetEnhet.right()
        } catch (e: WebClientResponseException) {
            log.error("Feil ved kall til Enhetsregisteret ved henting av enhet ${orgnrTilEnhet.verdi}", e)
            HentEnhetFeil.FeilVedKallTilEnhetsregisteret.left()
        } catch (e: DecodingException) {
            log.error("Feil ved dekoding av JSON ved henting av enhet ${orgnrTilEnhet.verdi}", e)
            HentEnhetFeil.FeilVedDekodingAvJson.left()
        }
    }


    open fun hentUnderenhet(orgnrTilUnderenhet: Orgnr): Either<HentUnderenhetFeil, Underenhet> {
        return try {
            val url = "${enhetsregisteretUrl}/underenheter/${orgnrTilUnderenhet.verdi}"
            val underenhet = webClient.get()
                .uri(url)
                .retrieve()
                .bodyToMono(UnderenhetResponseJson::class.java)
                .retryWhen(retrySpec)
                .block()!!
                .toDomain()
            if (!validerReturnertOrgnr(orgnrTilUnderenhet, underenhet.orgnr)) {
                return HentUnderenhetFeil.OrgnrMatcherIkke.left()
            }
            underenhet.right()
        } catch (e: WebClientResponseException) {
            if (e.statusCode.is5xxServerError) {
                log.error("Enhetsregisteret svarer ikke", e)
                return HentUnderenhetFeil.EnhetsregisteretSvarerIkke.left()
            } else {
                log.error("Feil ved kall til Enhetsregisteret ved henting av underenhet ${orgnrTilUnderenhet.verdi}", e)
                return HentUnderenhetFeil.FeilVedKallTilEnhetsregisteret.left()
            }
        } catch (e: DecodingException) {
            log.error("Feil ved dekoding av JSON ved henting av underenhet ${orgnrTilUnderenhet.verdi}", e)
            return HentUnderenhetFeil.FeilVedDekodingAvJson.left()
        }
    }


    private fun validerReturnertOrgnr(
        opprinneligOrgnr: Orgnr,
        returnertOrgnr: Orgnr
    ): Boolean = if (opprinneligOrgnr != returnertOrgnr) {
        log.error(
            "Orgnr hentet fra Enhetsregisteret samsvarer ikke med det medsendte orgnr. Request: ${opprinneligOrgnr.verdi}, response: ${returnertOrgnr.verdi}"
        )
        false
    } else true

    open fun healthcheck(): HttpStatusCode {
        return try {
            webClient.get()
                .uri(enhetsregisteretUrl)
                .retrieve()
                .toBodilessEntity()
                .block()!!
                .statusCode
        } catch (e: WebClientResponseException) {
            e.statusCode
        }
    }

    sealed class HentEnhetFeil {
        object FeilVedKallTilEnhetsregisteret : HentEnhetFeil()
        object FeilVedDekodingAvJson : HentEnhetFeil()
        object OrgnrMatcherIkke : HentEnhetFeil()
    }

    sealed class HentUnderenhetFeil {
        object EnhetsregisteretSvarerIkke : HentUnderenhetFeil()
        object FeilVedDekodingAvJson : HentUnderenhetFeil()
        object FeilVedKallTilEnhetsregisteret : HentUnderenhetFeil()
        object OrgnrMatcherIkke : HentUnderenhetFeil()
    }
}